addons/maintenance/models/maintenance.py

```diff
@@ -326,11 +326,12 @@ def write(self, vals):
         # the stage (stage_id) of the Maintenance Request changes.
         if vals and 'kanban_state' not in vals and 'stage_id' in vals:
             vals['kanban_state'] = 'normal'
-        if 'stage_id' in vals and self.maintenance_type == 'preventive' and self.recurring_maintenance and self.env['maintenance.stage'].browse(vals['stage_id']).done:
-            schedule_date = self.schedule_date or fields.Datetime.now()
-            schedule_date += relativedelta(**{f"{self.repeat_unit}s": self.repeat_interval})
-            if self.repeat_type == 'forever' or schedule_date.date() <= self.repeat_until:
-                self.copy({'schedule_date': schedule_date, 'stage_id': self._default_stage().id})
+        for request in self:
+            if 'stage_id' in vals and request.maintenance_type == 'preventive' and request.recurring_maintenance and request.env['maintenance.stage'].browse(vals['stage_id']).done:
+                schedule_date = request.schedule_date or fields.Datetime.now()
```

Comment:
Little NitPick but I think it is best to not call `now` in a loop. If you write on 1000 records, you might end up with different results since python did not froze time for the transaction.
```suggestion
        now = fields.Datetime.now()
        for request in self:
            if 'stage_id' in vals and request.maintenance_type == 'preventive' and request.recurring_maintenance and request.env['maintenance.stage'].browse(vals['stage_id']).done:
                schedule_date = request.schedule_date or now
```